### PR TITLE
[proxy] Improve `no_proxy` matching logic

### PR DESCRIPTION
**Pull request by olivielpeau**
### What does this PR do?

Before this commit, the `no_proxy` entries would only match hosts exactly. This adds support for a lot of cases where `no_proxy` entries should match the request's host, for example:
* a no_proxy entry that's a suffix of the target host now matches
* the port of the target host is now cleaned up before matching it with
  no_proxy entries
* [logic that might be golang-specific] localhost and loopback addresses
  never use proxies

(see the tests that were added for examples of target hosts that would stop using the configured http/https proxies)

### Motivation

Make the behavior of the `no_proxy` setting closer to what other programs (typically, `curl`) would do.

### Additional Notes

See second commit for the logic changes, the first commit only moves code around.

The code itself is mostly copied from `net/http/transport.go` of go's stdlib, with a few minor tweaks for our needs (see inline comments). Unfortunately the logic there isn't usable through exported functions.

### CODE REVIEW

File: pkg/util/http.go [hunk 1]
```diff
@@ -14,3 +26,160 @@ var apiKeyRegExp = regexp.MustCompile("api_key=*\\w+(\\w{5})")
...
+
+var portMap = map[string]string{
+	"http":   "80",
+	"https":  "443",
+	"socks5": "1080",
+}
+
+// canonicalAddr returns url.Host but always with a ":port" suffix.
+// Copied from `net/http/transport.go`, modified to remove the idna conversions.
+func canonicalAddr(url *url.URL) string {
```
Comment (truthbk) [inline pkg/util/http.go]: Note: Although to practical effects this code covers all use cases for us, I think it's incomplete strictly speaking. I believe with a SOCKS proxy (which we don't really support), any protocol really could be proxied so we could get a URI for something other than `http`, `https` and then we'd return `foo:0` which would possibly break. I don't think this affects us.

File: pkg/util/http.go [hunk 2]
```diff
@@ -14,3 +26,160 @@ var apiKeyRegExp = regexp.MustCompile("api_key=*\\w+(\\w{5})")
...
+		}
+
+		// check proxy by scheme
+		confProxy := ""
+		if r.URL.Scheme == "http" {
+			confProxy = p.HTTP
+		} else if r.URL.Scheme == "https" {
+			confProxy = p.HTTPS
+		} else {
+			log.Warnf("Proxy configuration do not support scheme '%s'", r.URL.Scheme)
```
Comment (truthbk) [inline pkg/util/http.go]: typo: "configuration _does_ not support scheme"

**Review by truthbk — DISMISSED** [review]
Looks 👌 just a small typo!

File: pkg/util/http.go (hunk 1 above)
Comment (olivielpeau) [reply pkg/util/http.go]: thanks for pointing that out, indeed it doesn't affect us so I'll leave it as-is

File: pkg/util/http.go [hunk 3]
```diff
@@ -14,3 +26,160 @@ var apiKeyRegExp = regexp.MustCompile("api_key=*\\w+(\\w{5})")
...
+			return false, "loopback"
+		}
+	}
+
+	addr = strings.ToLower(strings.TrimSpace(addr))
+	if hasPort(addr) {
+		addr = addr[:strings.LastIndex(addr, ":")]
+	}
+
+	for _, p := range noProxy {
```
Comment (hush-hush) [inline pkg/util/http.go]: Most of the post-processing on the `no_proxy` config could be done here: https://github.com/DataDog/datadog-agent/blob/master/pkg/config/config.go#L346

For each transaction we're going to check the proxy, the function should be as quick as possible.

File: pkg/util/http.go [hunk 4]
```diff
@@ -14,3 +26,160 @@ var apiKeyRegExp = regexp.MustCompile("api_key=*\\w+(\\w{5})")
...
+		}
+
+		// check proxy by scheme
+		confProxy := ""
+		if r.URL.Scheme == "http" {
+			confProxy = p.HTTP
+		} else if r.URL.Scheme == "https" {
+			confProxy = p.HTTPS
+		} else {
+			log.Warnf("Proxy configuration does not support scheme '%s'", r.URL.Scheme)
```
Comment (hush-hush) [inline pkg/util/http.go]: We could return where and avoid the if line 60.

File: pkg/util/http.go [hunk 5]
```diff
@@ -14,3 +26,160 @@ var apiKeyRegExp = regexp.MustCompile("api_key=*\\w+(\\w{5})")
...
+		return false, "localhost"
+	}
+	if ip := net.ParseIP(host); ip != nil {
+		if ip.IsLoopback() {
+			return false, "loopback"
+		}
+	}
+
+	addr = strings.ToLower(strings.TrimSpace(addr))
+	if hasPort(addr) {
```
Comment (hush-hush) [inline pkg/util/http.go]: Why do we use `canonicalAddr` (which always add the port) if we remove it here (we also remove the port from the `no_proxy` list entries) ?

**Review by hush-hush — CHANGES_REQUESTED** [review]

File: pkg/util/http.go [hunk 6]
```diff
@@ -14,3 +26,160 @@ var apiKeyRegExp = regexp.MustCompile("api_key=*\\w+(\\w{5})")
...
+}
+
+// useProxy reports whether requests to addr should use a proxy,
+// according to the noProxy entries, and if a proxy should be used it
+// returns the matching noProxy entry.
+// addr is always a canonicalAddr with a host and port.
+// Copied from `net/http/transport.go`, modified to use the passed noProxy values
+// instead of pulling directly from the no_proxy env var, and to return the
+// matched noProxy entry.
+func useProxy(addr string, noProxy []string) (matches bool, matchingNoProxy string) {
```
Comment (hush-hush) [inline pkg/util/http.go]: When setting the environment variable `HTTP_PROXY` all call to the cloud providers metadata endpoint are going through the proxy.

I think this function should white-list the those IP:
- `169.254.169.254` for ec2, gce and azure
- `169.254.170.2` for ecs (maybe not this one ?)

**Comment by hush-hush** [comment]
@olivielpeau should we close this ?

**Comment by olivielpeau** [comment]
Yes, even though I think the purpose of the PR is still valid, the PR would require being completely re-implemented to work well with the previous planned proxy logic changes. Closing

**Pull request closed without merging**